src/auth/mod.rs

```
@@ -0,0 +1,174 @@
+use hkdf::{Hkdf, InvalidLength};
+use hmac::{
+    crypto_mac::{InvalidKeyLength, MacError},
+    Hmac, Mac,
+};
+use ring;
+use serde_json::{self, Error as JsonError};
+use sha2::Sha256;
+use time::Duration;
+
+use settings::Settings;
+
+#[derive(Debug, Deserialize, PartialEq, Serialize)]
+pub struct HawkPayload {
+    pub expires: f64,
+    pub node: String,
+    pub salt: String,
+    pub uid: u64,
+}
+
+impl HawkPayload {
+    fn new(
+        header: &str,
+        method: &str,
+        path: &str,
+        host: &str,
+        port: u16,
+        settings: &Settings,
+        expiry: u64,
+    ) -> AuthResult<HawkPayload> {
+        if &header[0..5] != "Hawk " {
+            return Err(AuthError);
+        }
+
+        let header: HawkHeader = header[5..].parse()?;
+        let id = header.id.as_ref().ok_or(AuthError)?;
+
+        let payload = HawkPayload::extract_and_validate(id, &settings.master_token_secret, expiry)?;
+
+        let token_secret = hkdf_expand_32(
+            format!("services.mozilla.com/tokenlib/v1/derive/{}", id).as_bytes(),
+            Some(payload.salt.as_bytes()),
+            &settings.master_token_secret,
+        )?;
+        let token_secret = base64::encode_config(&token_secret, base64::URL_SAFE);
+
+        let request = RequestBuilder::new(method, host, port, path).request();
+        if request.validate_header(
+            &header,
+            &Key::new(token_secret.as_bytes(), &ring::digest::SHA256),
+            // Allow plenty of leeway for clock skew, because
+            // client timestamps tend to be all over the shop
+            Duration::weeks(52),
+        ) {
+            Ok(payload)
+        } else {
+            Err(AuthError)
+        }
+    }
+
+    fn extract_and_validate(
+        id: &str,
+        master_secret: &[u8],
+        expiry: u64,
+    ) -> AuthResult<HawkPayload> {
+        let decoded_id = base64::decode_config(id, base64::URL_SAFE)?;
+        let payload_length = decoded_id.len() - 32;
```

could decoded_id be < 32 here?

---

I was about to say "no", but then I realised I really mean "not if it's a valid header". But of course we still need to fail gracefully for invalid input so I'll add a length check before this step.